libs/meta/tests/unit/callable_traits_tests.cpp

```diff
@@ -0,0 +1,253 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/macros.hpp"
+#include "meta/callable_traits.hpp"
+#include "same.hpp"
+
+#include "gmock/gmock.h"
+
+#include <cstddef>
+#include <string>
+#include <tuple>
+#include <type_traits>
+#include <vector>
+
+using namespace ::testing;
+
+namespace fetch {
+namespace meta {
+
+namespace {
+
+using EmptyArgs = std::tuple<>;
+#define ARGS std::size_t const &, int, std::vector<std::string> &
+using Args         = std::tuple<ARGS>;
+using Void         = void;
+using NonVoid      = std::string;
+using NoOwningType = void;
+
+#define ASSERTIONS(callable, return_type, args_tuple_type, owning_type)      \
+  {                                                                          \
+    using ClassType = owning_type;                                           \
+    using Callable  = decltype(callable);                                    \
+                                                                             \
+    using Traits = CallableTraits<Callable>;                                 \
+    static_assert(Same<return_type, Traits::ReturnType>(), "");              \
+    static_assert(Same<args_tuple_type, Traits::ArgsTupleType>(), "");       \
+    static_assert(Same<ClassType, Traits::OwningType>(), "");                \
+                                                                             \
+    ASSERT_EQ(Traits::arg_count(), std::tuple_size<args_tuple_type>::value); \
+                                                                             \
+    auto const callable_is_void = Same<return_type, Void>();                 \
+    ASSERT_EQ(Traits::is_void(), callable_is_void);                          \
+  }
+
+Void free_function_void_no_args()
+{}
+
+NonVoid free_function_nonvoid_no_args()
+{
+  return "";
+}
+
+Void free_function_void(ARGS)
+{}
+
+NonVoid free_function_nonvoid(ARGS)
+{
+  return "";
+}
+
+class TestFunctions
+{
+public:
+  // static member functions
+  static Void static_member_function_void_no_args()
+  {}
+
+  static NonVoid static_member_function_nonvoid_no_args()
+  {
+    return "";
```

Comment:
`return NonVoid()`? Since it is typedefed anyway?

Reply:
Right, no point to reveal that it's a string by typing `""` if I had hidden `std::string` behind an alias to begin with. That said, I'll go with `{}` to keep things DRY.